### PR TITLE
make coarsen reductions consistent with reductions on other classes

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Tests added
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

This PR uses `inject_reduce_methods` to inject reduction methods into the Coarsen classes. So now we can do `coarsen.count()` and pass `skipna` down to the reduction methods.

### CODE REVIEW

**Comment by max-sixty** [comment]
Great idea, there's lots of room to centralize some of these!

**Comment by dcherian** [comment]
@fujisoup this could use a review if you have time.

File: xarray/core/variable.py [hunk 1]
```diff
@@ -1843,11 +1843,11 @@ def coarsen(self, windows, func, boundary="exact", side="left"):
             func = getattr(duck_array_ops, name, None)
             if func is None:
                 raise NameError(f"{name} is not a valid method.")
-        return type(self)(self.dims, func(reshaped, axis=axes), self._attrs)
+        return type(self)(self.dims, func(reshaped, axis=axes, **kwargs), self._attrs)
```
Comment (fujiisoup) [inline xarray/core/variable.py]: Probably now we can use `_replace` method?

**Comment by dcherian** [comment]
> I don't remember why I added a separate injection function...

:+1:

> Can you add some tests for coarsen.count() and skipna in the reduction methods?

There is a `count` test in `test_variable`. I've added `sum(skipna=...)` tests there too now.

File: doc/whats-new.rst [hunk 2]
```diff
@@ -85,6 +85,9 @@ New Features
   invoked. (:issue:`3378`, :pull:`3446`, :pull:`3515`)
   By `Deepak Cherian <https://github.com/dcherian>`_ and
   `Guido Imperiale <https://github.com/crusaderky>`_.
+- Added the :py:meth:`count` reduction method to both :py:class:`DatasetCoarsen`
+  and :py:class:`DataArrayCoarsen` objects. (:pull:`3500`)
+  By `Deepak Cherian <https://github.com/dcherian/>`_
```
Comment (keewis) [inline doc/whats-new.rst]: now that #3516 is merged, this should be
```suggestion
  By `Deepak Cherian <https://github.com/dcherian>`_
```

Comment (dcherian) [reply doc/whats-new.rst]: thanks!

**Comment by dcherian** [comment]
> Probably I wondered if bottleneck works with multiple reduction axes.

I'll check this before merging.

**Comment by dcherian** [comment]
In the case of multiple reduction axes i.e. `isinstance(axis, tuple) is True` we don't use the bottleneck method so things should be fine.

https://github.com/pydata/xarray/blob/ed05f9862622b00f40f7b9b99ccdb0ab3766ff0f/xarray/core/nputils.py#L210-L218

**Pull request merged**